compiler/rustc_parse/src/parser/item.rs

```diff
@@ -604,9 +609,17 @@ impl<'a> Parser<'a> {
                     // This notably includes paths passed through `ty` macro fragments (#46438).
                     TyKind::Path(None, path) => path,
                     _ => {
-                        self.sess.emit_err(errors::ExpectedTraitInTraitImplFoundType {
-                            span: ty_first.span,
-                        });
+                        if self.may_recover() && let Some(extra_impl_kw) = extra_impl_kw {
```

Comment:
I actually liked the previous approach. You probably need to check that the `impl Trait` only has one bound (isn't `impl A + B`) and you can probably use the span from the beginning of the ty to the beginning of the first bound for the removal span.